### PR TITLE
[bugfix] QNSPSAOptimizer improperly handles optimization on lightning.qubit

### PR DESCRIPTION
**Pull request by PietropaoloFrisoni**
**Context:** The `QNSPSAOptimizer` improperly handles optimization on devices that don't follow the new API design.

**Description of the Change:** In the workflow of the `step_and_cost` method of `QNSPSAOptimizer`, the code now checks whether the device follows the new API guidelines before calling the `preprocess` function (which is not present in the old device interface). If this is not the case, the code simply calls the `qml.execute` function.

**Benefits:** Now, `qml.QNSPSAOptimizer` can handle legacy devices that do not follow the new API design, such as `default.qubit.legacy`. Therefore, the following code now works:

```
from pennylane import numpy as np
import pennylane as qml

dev = qml.device("default.qubit.legacy", wires=2)

@qml.qnode(dev)
def cost(params):
    qml.RX(params[0], wires=0)
    qml.CRY(params[1], wires=[0, 1])
    return qml.expval(qml.Z(0) @ qml.Z(1))

opt = qml.QNSPSAOptimizer(stepsize=5e-2)
params = np.random.rand(2)

for _ in range(10):
    params, loss = opt.step_and_cost(cost, params)
```

**Possible Drawbacks:** None that I can think of.

**Related GitHub Issues:** #5460

[sc-60727], [sc-60328]

### CODE REVIEW

File: tests/optimize/test_qnspsa.py
Comment (PietropaoloFrisoni) [inline tests/optimize/test_qnspsa.py:1]: All tests in this file relying on `default.qubit` as a device, now work with `default.qubit.legacy` as well. Still, I just inserted the explicit check here so as not to overkill the test workflow.

**Review by albi3ro — APPROVED** [review]

**Review by isaacdevlugt — APPROVED** [review]
Looks good to me! Can confirm that the code example provided in the context works for me.

**Comment by trbromley** [comment]
Thanks @PietropaoloFrisoni! Should https://github.com/PennyLaneAI/pennylane/issues/5460 now be closed?

**Comment by PietropaoloFrisoni** [comment]
@trbromley Done. Thanks for the reminder!

**Pull request merged**